server/jetstream_test.go

```diff
@@ -19284,3 +19284,132 @@ func TestJetStreamConsumerEOFBugNewFileStore(t *testing.T) {
 	_, err = sub.NextMsg(time.Second)
 	require_NoError(t, err)
 }
+
+func TestJetStreamSubjectBasedFilteredConsumers(t *testing.T) {
+	conf := createConfFile(t, []byte(`
+		listen: 127.0.0.1:-1
+		jetstream: {max_mem_store: 64GB, max_file_store: 10TB}
+		accounts: {
+			A: {
+				jetstream: enabled
+				users: [ {
+					user: u,
+					password: p
+					permissions {
+						publish {
+							allow: [
+								'ID.>',
+								'$JS.API.INFO',
+								'$JS.API.STREAM.>',
+								'$JS.API.CONSUMER.INFO.>',
+								'$JS.API.CONSUMER.CREATE.TEST.VIN-xxx.ID.foo.>', # Only allow ID.foo.
+							]
+							deny: [ '$JS.API.CONSUMER.CREATE.*', '$JS.API.CONSUMER.DURABLE.CREATE.*.*']
+						}
+					}
+				} ]
+			},
+		}
+	`))
+	defer removeFile(t, conf)
+
+	s, _ := RunServerWithConfig(conf)
+	if config := s.JetStreamConfig(); config != nil {
+		defer removeDir(t, config.StoreDir)
+	}
+	defer s.Shutdown()
+
+	nc, js := jsClientConnect(t, s, nats.UserInfo("u", "p"), nats.ErrorHandler(noOpErrHandler))
+	defer nc.Close()
+
+	_, err := js.AddStream(&nats.StreamConfig{
+		Name:     "TEST",
+		Subjects: []string{"ID.*.*"},
+	})
+	require_NoError(t, err)
+
+	for i := 0; i < 100; i++ {
+		js.Publish(fmt.Sprintf("ID.foo.%d", i*3), nil)
+		js.Publish(fmt.Sprintf("ID.bar.%d", i*3+1), nil)
+		js.Publish(fmt.Sprintf("ID.baz.%d", i*3+2), nil)
+	}
+	si, err := js.StreamInfo("TEST")
+	require_NoError(t, err)
+	require_True(t, si.State.Msgs == 300)
+
+	// Trying to create a consumer with non filtered API should fail.
+	js, err = nc.JetStream(nats.MaxWait(200 * time.Millisecond))
+	require_NoError(t, err)
+
+	_, err = js.SubscribeSync("ID.foo.*")
+	require_Error(t, err, nats.ErrTimeout)
+
+	js, err = nc.JetStream(nats.MaxWait(200 * time.Millisecond))
```

Comment:
Curious as to why you need to recreate the JS context here? (since it was created few lines above).

Reply:
Probably don't since it is done above, will remove.

Reply:
The subscribe is supposed to fail, and would by default take 10s, hence why I recreate but could also just do for that one call I think..